File: frontend/src/features/purchase/types/purchase.ts
export interface Supplier {
  id: string;
  name: string;
  contact_person?: string;
  phone?: string;
  email?: string;
  address?: string;
  tax_number?: string;
  credit_limit: number;
  opening_balance: number;
  current_balance: number;
  is_active: boolean;
}

export interface SupplierLedger {
  id: string;
  transaction_date: string;
  transaction_type: string;
  reference_id: string;
  debit: number;
  credit: number;
  balance_after: number;
  notes?: string;
}

export interface POItem {
  id: string;
  medicine_id: string;
  medicine_name?: string; // Optional field populated by frontend join or backend
  quantity_ordered: number;
  quantity_received: number;
  unit_price: number;
}

export interface PurchaseOrder {
  id: string;
  order_number: string;
  supplier_id: string;
  supplier_name?: string; // Frontend helper
  expected_delivery_date?: string;
  total_amount: number;
  status: string; // Draft, Approved, Partially Received, Completed, Cancelled
  items: POItem[];
  created_at?: string;
  updated_at?: string;
}

export interface GRN {
  id: string;
  grn_number: string;
  po_id: string;
  supplier_id: string;
  received_date?: string;
  total_amount: number;
  status: string;
}

export interface InvoiceItem {
  medicine_id: string;
  medicine_name: string;
  batch_number: string;
  quantity: number;
  unit_price: number;
  total_price: number;
}

export interface PurchaseInvoice {
  id: string;
  invoice_number?: string;
  grn_id: string;
  supplier_id: string;
  invoice_date?: string;
  due_date?: string;
  total_amount: number;
  tax_amount: number;
  amount_paid: number;
  status: string; // Draft, Unpaid, Partially Paid, Paid, Cancelled
  items?: InvoiceItem[];
}

export interface SupplierPayment {
  id: string;
  supplier_id: string;
  invoice_id?: string;
  amount: number;
  payment_method: string;
  reference_number?: string;
  notes?: string;
  payment_date: string;
}

// Payloads
export interface CreateSupplierPayload {
  name: string;
  contact_person?: string;
  phone?: string;
  email?: string;
  address?: string;
  tax_number?: string;
  credit_limit: number;
  opening_balance: number;
  is_active: boolean;
}

export interface CreatePOPayload {
  supplier_id: string;
  expected_delivery_date?: string;
  total_amount: number;
  items: {
    medicine_id: string;
    quantity_ordered: number;
    unit_price: number;
  }[];
}

export interface GRNItemPayload {
  po_item_id: string;
  medicine_id: string;
  batch_number: string;
  manufacturing_date?: string;
  expiry_date: string;
  purchase_price: number;
  selling_price: number;
  quantity_received: number;
}

export interface CreateGRNPayload {
  po_id: string;
  supplier_id: string;
  received_date?: string;
  total_amount: number;
  items: GRNItemPayload[];
}

export interface CreateInvoicePayload {
  grn_id: string;
  supplier_id: string;
  invoice_date?: string;
  due_date?: string;
  total_amount: number;
  tax_amount: number;
  amount_paid: number;
}

export interface CreatePaymentPayload {
  supplier_id: string;
  invoice_id?: string;
  amount: number;
  payment_method: string;
  reference_number?: string;
  notes?: string;
}

// Enterprise Types

export interface PurchaseRequestItem {
  id: string;
  medicine_id: string;
  medicine_name?: string;  // Resolved from medicines table by backend
  quantity_requested: number;
  quantity_approved?: number;
  remarks?: string;
}

export interface PurchaseRequest {
  id: string;
  request_number: string;
  branch_id?: string;
  requested_by?: string;
  requested_by_name?: string;  // Resolved from users table by backend
  request_date?: string;
  required_date?: string;
  status: string; // Draft, Submitted, Approved, PO Created, Rejected
  priority: string;
  remarks?: string;
  items: PurchaseRequestItem[];
}

export interface SupplierScorecard {
  supplier_id: string;
  supplier_name: string;
  total_orders: number;
  on_time_delivery_pct: number;
  quality_score: number;
  return_rate: number;
  avg_lead_time_days: number;
  last_purchase_price?: number;
  overall_score: number;
}

export interface QuotationComparisonItem {
  medicine_id: string;
  medicine_name: string;
  unit_price: number;
  discount_percentage: number;
  tax_percentage: number;
  lead_time_days: number;
  line_total: number;
  is_lowest_price: boolean;
  is_fastest: boolean;
}

export interface QuotationComparisonEntry {
  quotation_id: string;
  quotation_number: string;
  supplier_id: string;
  supplier_name: string;
  total_amount: number;
  subtotal: number;
  discount_amount: number;
  tax_amount: number;
  valid_until?: string;
  currency: string;
  payment_terms?: string;
  delivery_terms?: string;
  status: string;
  items: QuotationComparisonItem[];
  scorecard?: SupplierScorecard;
  is_lowest_cost: boolean;
  is_fastest_delivery: boolean;
  is_best_overall: boolean;
  is_highest_rated: boolean;
}

export interface QuotationComparisonResponse {
  request_id: string;
  request_number?: string;
  medicine_ids: string[];
  quotations: QuotationComparisonEntry[];
}

export interface PurchaseQuotationItem {
  id?: string;
  medicine_id: string;
  medicine_name?: string;
  quantity: number;
  unit_price: number;
  discount_percentage: number;
  tax_percentage: number;
  lead_time_days: number;
  line_total?: number;
  moq?: number;
  brand?: string;
  manufacturer?: string;
  batch_number?: string;
  expiry_date?: string;
  line_notes?: string;
}

export interface PurchaseQuotation {
  id: string;
  quotation_number: string;
  request_id?: string;
  request_number?: string;
  supplier_id: string;
  supplier_name?: string;
  branch_id?: string;
  warehouse_id?: string;
  quotation_date?: string;
  valid_until?: string;
  currency: string;
  status: string;
  total_amount: number;
  subtotal: number;
  discount_amount: number;
  tax_amount: number;
  payment_terms?: string;
  delivery_terms?: string;
  warranty?: string;
  remarks?: string;
  attachment_url?: string;
  supplier_score?: number;
  items: PurchaseQuotationItem[];
}

export interface PurchaseApproval {
  id: string;
  po_id: string;
  approver_id: string;
  level: number;
  status: string; // Pending, Approved, Rejected
  comments?: string;
}

export interface PurchaseReceivingItem {
  id: string;
  po_item_id: string;
  medicine_id: string;
  batch_number: string;
  expiry_date?: string;
  manufacturing_date?: string;
  received_quantity: number;
  accepted_quantity: number;
  rejected_quantity: number;
  rejection_reason?: string;
}

export interface PurchaseReceiving {
  id: string;
  receiving_number: string;
  po_id: string;
  warehouse_id?: string;
  received_by_id: string;
  status: string; // Draft, Completed
  items: PurchaseReceivingItem[];
}

export interface CreatePurchaseRequestPayload {
  priority: string;
  status: string; // 'Draft' or 'Submitted'
  remarks?: string;
  required_date?: string;
  items: {
    medicine_id: string;
    quantity_requested: number;
    quantity_approved?: number;
    remarks?: string;
  }[];
}

export interface CreatePurchaseQuotationPayload {
  supplier_id: string;
  request_id?: string;
  branch_id?: string;
  warehouse_id?: string;
  quotation_date?: string;
  valid_until?: string;
  currency?: string;
  status?: string;
  total_amount?: number;
  subtotal?: number;
  discount_amount?: number;
  tax_amount?: number;
  payment_terms?: string;
  delivery_terms?: string;
  warranty?: string;
  remarks?: string;
  attachment_url?: string;
  items: PurchaseQuotationItem[];
}

export interface QuotationStatusUpdate {
  status: string;
  remarks?: string;
}

export interface UpdatePurchaseQuotationPayload extends Partial<CreatePurchaseQuotationPayload> {}

export interface CreatePurchaseApprovalPayload {
  po_id: string;
  status: string;
  comments?: string;
}

export interface CreatePurchaseReceivingPayload {
  po_id: string;
  warehouse_id?: string;
  items: { 
    po_item_id: string; 
    medicine_id: string; 
    batch_number: string; 
    expiry_date?: string; 
    manufacturing_date?: string; 
    received_quantity: number; 
    accepted_quantity: number; 
    rejected_quantity: number; 
    rejection_reason?: string; 
  }[];
}


export interface PurchaseApprovalMatrix {
  id: string;
  level: number;
  role_name: string;
  amount_threshold: number;
}

export interface PurchaseTimeline {
  id: string;
  reference_id: string;
  reference_type: string;
  action: string;
  user_id?: string;
  remarks?: string;
  timestamp: string;
}

export interface CreatePurchaseApprovalMatrixPayload {
  level: number;
  role_name: string;
  amount_threshold: number;
}

export interface PurchaseApprovalRequestPayload {
  status: string;
  remarks?: string;
}

File: backend/models/purchase.py
from sqlalchemy import Column, String, Boolean, Float, Integer, ForeignKey, Text, Date, DateTime
from sqlalchemy.orm import relationship
from typing import Optional
from datetime import datetime
from sqlalchemy.orm import relationship
from typing import Optional
from .base import BaseModel

class Supplier(BaseModel):
    __tablename__ = "suppliers"

    name = Column(String(255), nullable=False)
    contact_person = Column(String(255))
    phone = Column(String(50))
    email = Column(String(255))
    address = Column(Text)
    region_name = Column(String(100), nullable=True)
    tax_number = Column(String(100))
    credit_limit = Column(Float, default=0.0)
    opening_balance = Column(Float, default=0.0)
    current_balance = Column(Float, default=0.0)
    is_active = Column(Boolean, default=True)

class PurchaseOrder(BaseModel):
    __tablename__ = "purchase_orders"
    
    order_number = Column(String(100), unique=True, index=True)
    supplier_id = Column(String(36), ForeignKey("suppliers.id"))
    branch_id = Column(String(36), ForeignKey("branches.id"))
    
    expected_delivery_date = Column(Date)
    # Draft, Submitted, Pending Approval, Approved, Partially Received, Completed, Cancelled
    status = Column(String(50), default="Draft")
    total_amount = Column(Float, default=0.0)
    
    # Enterprise Fields
    warehouse_id = Column(String(36), ForeignKey("branch_warehouses.id"), nullable=True)
    approved_by = Column(String(36), ForeignKey("users.id"), nullable=True)
    approved_at = Column(DateTime, nullable=True)
    approval_status = Column(String(50), default="Pending") # Pending, Approved, Rejected
    delivery_status = Column(String(50), default="Pending") # Pending, Partial, Delivered
    payment_status = Column(String(50), default="Unpaid")   # Unpaid, Partial, Paid
    purchase_priority = Column(String(50), default="Normal") # Normal, High, Urgent
    purchase_source = Column(String(50), default="Manual")  # Manual, Auto, Request, Quotation
    reference_number = Column(String(100), nullable=True)
    remarks = Column(Text, nullable=True)
    
    items = relationship("POItem", back_populates="purchase_order")
    supplier = relationship("Supplier")

    @property
    def supplier_name(self) -> Optional[str]:
        return self.supplier.name if self.supplier else None

class POItem(BaseModel):
    __tablename__ = "po_items"
    
    po_id = Column(String(36), ForeignKey("purchase_orders.id"))
    medicine_id = Column(String(36), ForeignKey("medicines.id"))
    
    quantity_ordered = Column(Integer, nullable=False)
    quantity_received = Column(Integer, default=0)
    unit_price = Column(Float, default=0.0)
    
    # Enterprise Fields
    warehouse_id = Column(String(36), ForeignKey("branch_warehouses.id"), nullable=True)
    pending_quantity = Column(Integer, default=0)
    rejected_quantity = Column(Integer, default=0)
    bonus_quantity = Column(Integer, default=0)
    discount_percentage = Column(Float, default=0.0)
    tax_percentage = Column(Float, default=0.0)
    expiry_required = Column(Boolean, default=True)
    batch_required = Column(Boolean, default=True)
    
    purchase_order = relationship("PurchaseOrder", back_populates="items")
    medicine = relationship("Medicine", foreign_keys=[medicine_id], lazy="joined")

    @property
    def medicine_name(self) -> Optional[str]:
        return self.medicine.name if self.medicine else None


class GRN(BaseModel):
    __tablename__ = "grns"

    grn_number = Column(String(100), unique=True, index=True)
    po_id = Column(String(36), ForeignKey("purchase_orders.id"), nullable=True)
    supplier_id = Column(String(36), ForeignKey("suppliers.id"))
    branch_id = Column(String(36), ForeignKey("branches.id"))
    
    received_date = Column(Date)
    status = Column(String(50), default="Draft") # Draft, Confirmed
    total_amount = Column(Float, default=0.0)
    
    supplier = relationship("Supplier")

    @property
    def supplier_name(self) -> Optional[str]:
        return self.supplier.name if self.supplier else None
    
class PurchaseInvoice(BaseModel):
    __tablename__ = "purchase_invoices"

    invoice_number = Column(String(100), unique=True, index=True)
    grn_id = Column(String(36), ForeignKey("grns.id"))
    supplier_id = Column(String(36), ForeignKey("suppliers.id"))
    
    invoice_date = Column(Date)
    due_date = Column(Date)
    total_amount = Column(Float, default=0.0)
    tax_amount = Column(Float, default=0.0)
    amount_paid = Column(Float, default=0.0)
    
    status = Column(String(50), default="Unpaid") # Unpaid, Partial, Paid

class SupplierPayment(BaseModel):
    __tablename__ = "supplier_payments"
    
    supplier_id = Column(String(36), ForeignKey("suppliers.id"))
    branch_id = Column(String(36), ForeignKey("branches.id"))
    invoice_id = Column(String(36), ForeignKey("purchase_invoices.id"), nullable=True) # Optional if advance payment
    
    payment_date = Column(Date)
    amount = Column(Float, nullable=False)
    payment_method = Column(String(50)) # Cash, Bank Transfer, Cheque
    reference_number = Column(String(100))
    notes = Column(Text)

class PurchaseReturn(BaseModel):
    __tablename__ = "purchase_returns"
    
    return_number = Column(String(100), unique=True, index=True)
    po_id = Column(String(36), ForeignKey("purchase_orders.id"), nullable=True)
    grn_id = Column(String(36), ForeignKey("grns.id"), nullable=True)
    supplier_id = Column(String(36), ForeignKey("suppliers.id"))
    branch_id = Column(String(36), ForeignKey("branches.id"))
    
    return_date = Column(Date)
    total_amount = Column(Float, default=0.0)
    reason = Column(String(255))
    status = Column(String(50), default="Draft") # Draft, Completed

    items = relationship("PurchaseReturnItem", back_populates="purchase_return")
    supplier = relationship("Supplier")

class PurchaseReturnItem(BaseModel):
    __tablename__ = "purchase_return_items"

    purchase_return_id = Column(String(36), ForeignKey("purchase_returns.id"))
    medicine_id = Column(String(36), ForeignKey("medicines.id"))
    
    quantity_returned = Column(Integer, nullable=False, default=1)
    unit_price = Column(Float, default=0.0)
    
    purchase_return = relationship("PurchaseReturn", back_populates="items")
    medicine = relationship("Medicine", foreign_keys=[medicine_id], lazy="joined")
    
    @property
    def medicine_name(self) -> Optional[str]:
        return self.medicine.name if self.medicine else None

class SupplierLedger(BaseModel):
    __tablename__ = "supplier_ledger"
    
    supplier_id = Column(String(36), ForeignKey("suppliers.id"))
    branch_id = Column(String(36), ForeignKey("branches.id"), nullable=True)
    
    transaction_date = Column(Date)
    transaction_type = Column(String(50)) # Invoice, Payment, Return
    reference_id = Column(String(100)) # ID of Invoice, Payment, Return
    
    debit = Column(Float, default=0.0) # Increases supplier balance (we owe them) -> Invoices
    credit = Column(Float, default=0.0) # Decreases supplier balance -> Payments, Returns
    balance_after = Column(Float, nullable=False)
    
    notes = Column(Text)

class SupplierMedicinePrice(BaseModel):
    __tablename__ = "supplier_medicine_prices"

    supplier_id = Column(String(36), ForeignKey("suppliers.id"), nullable=False)
    medicine_id = Column(String(36), ForeignKey("medicines.id"), nullable=False)
    
    trade_price = Column(Float, nullable=False, default=0.0)
    exclusive_discount_percentage = Column(Float, default=0.0)
    bonus_scheme_threshold = Column(Integer, default=0) # e.g. buy 10 get 1 free
    delivery_lead_time_days = Column(Integer, default=1)
    
    supplier = relationship("Supplier")
    medicine = relationship("Medicine")

# =====================================================================
# Enterprise Purchase Models
# =====================================================================

class PurchaseApproval(BaseModel):
    __tablename__ = "purchase_approvals"
    
    po_id = Column(String(36), ForeignKey("purchase_orders.id"))
    approver_id = Column(String(36), ForeignKey("users.id"))
    level = Column(Integer, default=1)
    status = Column(String(50), default="Pending") # Pending, Approved, Rejected
    comments = Column(Text, nullable=True)
    approved_at = Column(DateTime, nullable=True)

class PurchaseReceiving(BaseModel):
    __tablename__ = "purchase_receivings"
    
    receiving_number = Column(String(100), unique=True, index=True)
    po_id = Column(String(36), ForeignKey("purchase_orders.id"), nullable=True)
    supplier_id = Column(String(36), ForeignKey("suppliers.id"))
    branch_id = Column(String(36), ForeignKey("branches.id"))
    warehouse_id = Column(String(36), ForeignKey("branch_warehouses.id"), nullable=True)
    
    received_date = Column(Date, default=datetime.utcnow)
    status = Column(String(50), default="Draft") # Draft, Confirmed
    total_amount = Column(Float, default=0.0)
    freight_charge = Column(Float, default=0.0)
    landed_cost_allocated = Column(Boolean, default=False)
    
    items = relationship("PurchaseReceivingItem", back_populates="receiving")

class PurchaseReceivingItem(BaseModel):
    __tablename__ = "purchase_receiving_items"
    
    receiving_id = Column(String(36), ForeignKey("purchase_receivings.id"))
    po_item_id = Column(String(36), ForeignKey("po_items.id"), nullable=True)
    medicine_id = Column(String(36), ForeignKey("medicines.id"))
    batch_id = Column(String(36), ForeignKey("batches.id"), nullable=True)
    
    quantity_received = Column(Integer, nullable=False)
    quantity_rejected = Column(Integer, default=0)
    bonus_quantity = Column(Integer, default=0)
    
    unit_price = Column(Float, default=0.0)
    discount_percentage = Column(Float, default=0.0)
    tax_percentage = Column(Float, default=0.0)
    landed_cost = Column(Float, default=0.0)
    
    receiving = relationship("PurchaseReceiving", back_populates="items")

class PurchaseRequest(BaseModel):
    __tablename__ = "purchase_requests"
    
    request_number = Column(String(100), unique=True, index=True)
    branch_id = Column(String(36), ForeignKey("branches.id"))
    requested_by = Column(String(36), ForeignKey("users.id"))
    
    request_date = Column(Date, default=datetime.utcnow)
    required_date = Column(Date, nullable=True)
    status = Column(String(50), default="Draft") # Draft, Submitted, Approved, PO Created, Rejected
    priority = Column(String(50), default="Normal")
    remarks = Column(Text, nullable=True)
    
    items = relationship("PurchaseRequestItem", back_populates="request")

class PurchaseRequestItem(BaseModel):
    __tablename__ = "purchase_request_items"
    
    request_id = Column(String(36), ForeignKey("purchase_requests.id"))
    medicine_id = Column(String(36), ForeignKey("medicines.id"))
    quantity_requested = Column(Integer, nullable=False)
    quantity_approved = Column(Integer, default=0)
    remarks = Column(Text, nullable=True)
    
    request = relationship("PurchaseRequest", back_populates="items")

class PurchaseQuotation(BaseModel):
    __tablename__ = "purchase_quotations"
    
    quotation_number = Column(String(100), unique=True, index=True)
    request_id = Column(String(36), ForeignKey("purchase_requests.id"), nullable=True)
    supplier_id = Column(String(36), ForeignKey("suppliers.id"))
    branch_id = Column(String(36), ForeignKey("branches.id"))
    
    valid_until = Column(Date, nullable=True)
    status = Column(String(50), default="Draft") # Draft, Submitted, Accepted, Rejected
    total_amount = Column(Float, default=0.0)
    remarks = Column(Text, nullable=True)
    
    items = relationship("PurchaseQuotationItem", back_populates="quotation")

class PurchaseQuotationItem(BaseModel):
    __tablename__ = "purchase_quotation_items"
    
    quotation_id = Column(String(36), ForeignKey("purchase_quotations.id"))
    medicine_id = Column(String(36), ForeignKey("medicines.id"))
    
    quantity = Column(Integer, nullable=False)
    unit_price = Column(Float, default=0.0)
    discount_percentage = Column(Float, default=0.0)
    tax_percentage = Column(Float, default=0.0)
    lead_time_days = Column(Integer, default=1)
    
    quotation = relationship("PurchaseQuotation", back_populates="items")

class SupplierPriceHistory(BaseModel):
    __tablename__ = "supplier_price_history"
    
    supplier_id = Column(String(36), ForeignKey("suppliers.id"))
    medicine_id = Column(String(36), ForeignKey("medicines.id"))
    
    effective_date = Column(Date, default=datetime.utcnow)
    trade_price = Column(Float, nullable=False)
    discount_percentage = Column(Float, default=0.0)
    source = Column(String(50)) # Quotation, PO, Manual

class SupplierContract(BaseModel):
    __tablename__ = "supplier_contracts"
    
    supplier_id = Column(String(36), ForeignKey("suppliers.id"))
    contract_number = Column(String(100), unique=True)
    
    start_date = Column(Date)
    end_date = Column(Date)
    status = Column(String(50), default="Active") # Active, Expired, Terminated
    minimum_order_value = Column(Float, default=0.0)
    rebate_percentage = Column(Float, default=0.0)
    document_url = Column(String(255), nullable=True)

class PurchaseAttachment(BaseModel):
    __tablename__ = "purchase_attachments"
    
    reference_type = Column(String(50)) # PO, Receiving, Quotation, Invoice
    reference_id = Column(String(36))
    
    file_name = Column(String(255))
    file_url = Column(String(1024))
    file_type = Column(String(50))
    uploaded_by = Column(String(36), ForeignKey("users.id"))
    uploaded_at = Column(DateTime, default=datetime.utcnow)

